src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Pages/BookmarkListItem.cs

```diff
@@ -0,0 +1,304 @@
+        OnPropertyChanged(args.PropertyName);
+
+    private static void BuildSpecificContextMenuItems(Classification classification, List<IContextItem> contextMenu)
+    {
+        // TODO: unify across all built-in extensions
+        var bookmarkTargetType = classification.Kind;
+
+        // TODO: add "Run as administrator" for executables/shortcuts
+        if (!classification.IsPlaceholder)
+        {
+            if (bookmarkTargetType == CommandKind.FileDocument && File.Exists(classification.Target))
+            {
+                contextMenu.Add(new CommandContextItem(new OpenWithCommand(classification.Input)));
+            }
+        }
+
+        string? directoryPath = null;
+        var targetPath = classification.Target;
+        switch (bookmarkTargetType)
+        {
+            case CommandKind.Directory:
+                directoryPath = targetPath;
+                contextMenu.Add(new CommandContextItem(new DirectoryPage(directoryPath))); // Browse
+                break;
+            case CommandKind.FileExecutable:
+            case CommandKind.FileDocument:
+            case CommandKind.Shortcut:
+            case CommandKind.InternetShortcut:
+                try
+                {
+                    directoryPath = Path.GetDirectoryName(targetPath);
+                }
+                catch
+                {
+                    // ignore any path parsing errors
+                }
+
+                break;
+            case CommandKind.WebUrl:
+            case CommandKind.Protocol:
+            case CommandKind.Aumid:
+            case CommandKind.PathCommand:
+            case CommandKind.Unknown:
+            default:
+                break;
+        }
+
+        // Add "Copy Path" or "Copy Address" command
+        if (!string.IsNullOrWhiteSpace(classification.Input))
+        {
+            var copyCommand = new CopyPathCommand(targetPath)
+            {
+                Name = bookmarkTargetType is CommandKind.WebUrl or CommandKind.Protocol
+                    ? Resources.bookmarks_copy_address_name
+                    : Resources.bookmarks_copy_path_name,
+                Icon = Icons.CopyPath,
+            };
+
+            contextMenu.Add(new CommandContextItem(copyCommand) { RequestedShortcut = KeyChords.CopyPath });
+        }
+
+        // Add "Open in Console" and "Show in Folder" commands if we have a valid directory path
+        if (!string.IsNullOrWhiteSpace(directoryPath) && Directory.Exists(directoryPath))
+        {
+            contextMenu.Add(new CommandContextItem(new ShowFileInFolderCommand(targetPath)) { RequestedShortcut = KeyChords.OpenFileLocation });
+            contextMenu.Add(new CommandContextItem(OpenInConsoleCommand.FromDirectory(directoryPath)) { RequestedShortcut = KeyChords.OpenInConsole });
+        }
+
+        if (!string.IsNullOrWhiteSpace(targetPath) && (File.Exists(targetPath) || Directory.Exists(targetPath)))
```

Comment:
possibly worth a follow-up: if you bookmark a path to a file share, and _that file share is incredibly slow_, then this will just block while the OS attempts to resolve that path.

(I have to work with some file shares that feel like they must be hosted on the moon or something)